packages/vite/src/node/plugins/workerImportMetaUrl.ts

```diff
@@ -0,0 +1,95 @@
+import type { ResolvedConfig } from '../config'
+import type { Plugin } from '../plugin'
+import { getAssetHash, fileToUrl } from './asset'
+import {
+  cleanUrl,
+  injectQuery,
+  multilineCommentsRE,
+  singlelineCommentsRE
+} from '../utils'
+import path from 'path'
+import { parseWorkerRequest, bundleWorkerEntry } from './worker'
+import { ENV_PUBLIC_PATH } from '../constants'
+import MagicString from 'magic-string'
+
+const WorkerFileId = 'worker_url_file'
+
+export function workerImportMetaUrlPlugin(config: ResolvedConfig): Plugin {
+  const isBuild = config.command === 'build'
+
+  return {
+    name: 'vite:worker-import-meta-url',
+
+    async transform(code, id, options) {
+      const query = parseWorkerRequest(id)
+      if (query && query[WorkerFileId] != null) {
+        return {
+          code: `import '${ENV_PUBLIC_PATH}'\n` + code
+        }
+      }
+      if (
+        (code.includes('new Worker') || code.includes('new ShareWorker')) &&
+        code.includes('new URL') &&
+        code.includes(`import.meta.url`)
+      ) {
+        const importMetaUrlRE =
+          /\bnew\s+(Worker|SharedWorker)\s*\(\s*(new\s+URL\s*\(\s*('[^']+'|"[^"]+"|`[^`]+`)\s*,\s*import\.meta\.url\s*\))/g
+        const noCommentsCode = code
+          .replace(multilineCommentsRE, (m) => ' '.repeat(m.length))
+          .replace(singlelineCommentsRE, (m) => ' '.repeat(m.length))
+        let match: RegExpExecArray | null
+        let s: MagicString | null = null
+        while ((match = importMetaUrlRE.exec(noCommentsCode))) {
+          const { 0: allExp, 2: exp, 3: rawUrl, index } = match
+
+          const urlIndex = allExp.indexOf(exp) + index
+
+          if (options?.ssr) {
+            this.error(
+              `\`new URL(url, import.meta.url)\` is not supported in SSR.`,
+              urlIndex
+            )
+          }
+
+          // potential dynamic template string
+          if (rawUrl[0] === '`' && /\$\{/.test(rawUrl)) {
+            this.error(
+              `\`new URL(url, import.meta.url)\` is not supported in dynamic template string.`,
+              urlIndex
+            )
+          }
+
+          if (!s) s = new MagicString(code)
```

Comment:
I think this is the same code as:

```suggestion
          s ||= new MagicString(code)
```